cmd/controller-manager/app/controller_manager.go

```diff
@@ -170,40 +176,44 @@ func Run(controllerManagerOptions *options.ControllerManagerServer) error {
 		panic("unreachable")
 	}
 
-	// TODO: leader election for this controller-manager
-	// id, err := os.Hostname()
-	// if err != nil {
-	// 	return err
-	// }
-	// rl := resourcelock.EndpointsLock{
-	// 	EndpointsMeta: v1.ObjectMeta{
-	// 		Namespace: "kube-system",
-	// 		Name:      "kube-controller-manager",
-	// 	},
-	// 	Client: leaderElectionClient,
-	// 	LockConfig: resourcelock.ResourceLockConfig{
-	// 		Identity:      id,
-	// 		EventRecorder: recorder,
-	// 	},
-	// }
+	if !controllerManagerOptions.LeaderElection.LeaderElect {
+		run(nil)
+		panic("unreachable")
+	}
+
+	// Identity used to distinguish between multiple cloud controller manager instances
+	id, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	// Lock required for leader election
+	rl := resourcelock.EndpointsLock{
+		EndpointsMeta: metav1.ObjectMeta{
+			Namespace: "service-catalog",
+			Name:      "service-catalog-controller-manager",
+		},
+		Client: leaderElectionClient,
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity:      id + "-sc-controller-controller",
+			EventRecorder: recorder,
+		},
+	}
 
-	// leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
-	// 	Lock:          &rl,
-	// 	LeaseDuration: controllerManagerOptions.LeaderElection.LeaseDuration.Duration,
-	// 	RenewDeadline: controllerManagerOptions.LeaderElection.RenewDeadline.Duration,
-	// 	RetryPeriod:   controllerManagerOptions.LeaderElection.RetryPeriod.Duration,
-	// 	Callbacks: leaderelection.LeaderCallbacks{
-	// 		OnStartedLeading: run,
-	// 		OnStoppedLeading: func() {
-	// 			glog.Fatalf("leaderelection lost")
-	// 		},
-	// 	},
-	// })
-	// panic("unreachable")
-
-	run(make(<-chan (struct{})))
-
-	return nil
+	// Try and become the leader and start cloud controller manager loops
+	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+		Lock:          &rl,
+		LeaseDuration: controllerManagerOptions.LeaderElection.LeaseDuration.Duration,
+		RenewDeadline: controllerManagerOptions.LeaderElection.RenewDeadline.Duration,
+		RetryPeriod:   controllerManagerOptions.LeaderElection.RetryPeriod.Duration,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				glog.Fatalf("leaderelection lost")
```

Comment:
I do not know the details. Checking my understanding here.

Are these callbacks only on level change? Going from leading -> not leading? I don't want the non leading controllers to be constantly killing themselves cause they aren't allowed to lead.